gas-tracker-frontend/persistence/gas-data-persistence.ts

```diff
@@ -1,32 +1,48 @@
 import { EventEmitter } from 'events';
-import { Observable, Subscriber } from 'rxjs';
+import { Connectable, connectable, Observable, ReplaySubject, Subject, Subscriber } from 'rxjs';
 
 import { GasDataKey } from '../gas-data-constants';
 import { GasData, GasDataPersistenceEvents, GasDatum } from '../types/gas-data.types';
 
 // This class combines a rxjs observable with an event emitter in order to provide a pub-sub interface.
+// I didn't use a Subject intentionally. tyvm.
 export class GasDataPersistence extends EventEmitter {
   private static instance: GasDataPersistence;
 
-  private gasData$: Observable<GasData> =
-    new Observable((subscriber: Subscriber<GasData>) => {
-      this.on(GasDataPersistenceEvents.addRecord, (gasDatum: GasDatum) => {
-        this.addGasDatum(gasDatum);
-        subscriber.next({ gasData: this.readGasData() });
-      });
+  public get getGasData$(): Connectable<GasData> {
+    return this.gasData$;
+  }
+
+  public static getInstance() {
+    if (GasDataPersistence.instance) {
+      return GasDataPersistence.instance;
+    }
 
-      this.on(GasDataPersistenceEvents.deleteRecordByTimestamp, async (timestamp) => {
-        this.deleteByTimestamp(timestamp)
-        subscriber.next({ gasData: this.readGasData() })
-      });
+    GasDataPersistence.instance = new this();
+    GasDataPersistence.instance.getGasData$.connect();
+    return GasDataPersistence.instance;
+  }
 
-      subscriber.next({ gasData: this.readGasData() })
+  // Called every time a subscription is created.
+  private initializeGasDataObservable(subscriber: Subscriber<GasData>) {
+    this.on(GasDataPersistenceEvents.addRecord, (gasDatum: GasDatum) => {
+      this.addGasDatum(gasDatum);
+      subscriber.next({ gasData: this.readGasData() });
     });
 
-  public get getGasData$() {
-    return this.gasData$;
+    this.on(GasDataPersistenceEvents.deleteRecordByTimestamp, async (timestamp) => {
+      this.deleteByTimestamp(timestamp);
+      subscriber.next({ gasData: this.readGasData() });
+    });
+
+    subscriber.next({ gasData: this.readGasData() });
   }
 
+  private gasData$: Connectable<GasData> =
+    connectable(new Observable(this.initializeGasDataObservable.bind(this)), {
+      connector: () => new ReplaySubject<GasData>(),
```

Comment:
Let's specify the buffer size since we only need to send the last record to new subscribers.